Update Nginx Plus plugin to use newer API (#8)
signalfx/collectd-nginx-plus: README.md

|:--------|:-----------|
| StatusHost | IP address or DNS of the NGINX+ instance to retrieve status information from. Defaults to `localhost`. |
| StatusPort | Port the NGINX+ status endpoint can be reached at. Defaults to `8080`. |

> | APIType | API Type of the Nginx-plus(`legacy` or `newer`). Detects the type by Default. (see below) |

Author: Here, by this I mean, if the host machine has Nginx-plus version `R11` or `R12` then they should mention `legacy`, if the nginx-plus version is between `R16` to the `latest`(both inclusive) then they should mention `newer` and if `R13` to `R15` then as per the Nginx config they should mention `newer` or `legacy`.
So, here `legacy` or `newer` means the range of versions as mentioned above, not the specific version, therefore any version between `R11` to the `latest` is supported by our plugin. 
And by default, the plugin will detect the Nginx-plus version and determine whether it is a `legacy` or `newer` type of API is running on the host's machine.

> | APIType | API Type of the Nginx-plus(`legacy` or `newer`). Detects the type by Default. (see below) |

Reviewer: I believe I understand what you mean and how it's implemented.  My request is that instead of "newer" or "legacy" we have a "legacy" option and then a configurable newer version (1, 2, 3, etc.).  This provides more configurability and would better allow us to set a newer default of a tested version like 3 that works across a range of possible nginx plus releases.  As is, there could be a breaking change w/ version 8 and we'd blindly move to it without being able to recover without implementing the feature I'm asking for.

> | APIType | API Type of the Nginx-plus(`legacy` or `newer`). Detects the type by Default. (see below) |

Author: Got your point, there was a misunderstanding from my side, sorry for that. This makes sense, I'll make the suggested changes.

> | APIType | API Type of the Nginx-plus(`legacy` or `newer`). Detects the type by Default. (see below) |

Author: @rmfitzpatrick, I have made the suggested changes, please review the same.

|:--------|:-----------|
| StatusHost | IP address or DNS of the NGINX+ instance to retrieve status information from. Defaults to `localhost`. |
| StatusPort | Port the NGINX+ status endpoint can be reached at. Defaults to `8080`. |

> | LegacyAPI | Indicate whether the NGINX+ is using `legacy` API or not. |

Reviewer: ```suggestion
| LegacyAPI | Indicate whether the NGINX+ is using `legacy` API or not. Defaults to `false`. |
```

Author: Hello @rmfitzpatrick,

As per the latest design(that exists in the current source code), we have removed the config option of  `LegacyAPI`.
We are currently driving the logic of determining `legacy` or `newer` APIs based on the single config option of `APIVersion`. A value of zero(0) in the `APIVersion` represents `legacy` APIs and a positive integer value in the `APIVersion` represents the API version of `newer` APIs.

Please let us know in case of any queries

> | APIVersion | API version to use for fetching data. Explicitly set this to `0` for the `legacy` versions of NGINX+. Defaults to `3`, which is for the `newer` API Versions. |

Reviewer: "newer" and "legacy" have no meaning after your recent changes:

```suggestion
| APIVersion | API version to use for fetching data. Explicitly set this to `0` for the legacy versions of NGINX+. Defaults to the versioned API of `3`. |
```

However, we should try to detect the version (including legacy) by default like was done earlier imo.  Defaulting to 3 for versioned cases makes sense to me.  Otherwise this is a breaking update for legacy API users.

Author: We brainstormed upon the point that you made above, and yes, providing compatibility with existing legacy API users is a better option. Following that aspect, we have made the below changes in the logical workflow of the code.

1. Both the APIVversion and APIBasePath are optional.
2. If APIVersion is provided, it'll be considered as versioned API NGINX+. Based on the provided APIBasePath (defaults to '/api' for versioned API) the further code workflow will be executed. If the legacy API is detected (on the server) in this case when APIVersion is provided by the end-user, then the code will exit with a proper error message asking the user to remove the APIVersion from the input configuration.
3. If APIVersion is not provided, the code will try to detect the NGINX+ API (versioned or legacy) on the server. If legacy API is detected, based on the provided APIBasePath (defaults to '/status' for legacy API) the further code workflow will be executed. If versioned API is detected, based on the provided APIBasePath (defaults to '/api' for versioned API) the further code workflow will be executed by taking the default APIVersion of 3.
I have added the same in the readme and also updated the test cases accordingly.
Does this sound reasonable?
